### DISCUSSION TITLE
Bug - Optional parameter with other parameters using params not working properly

### DISCUSSION THREAD

**Discussion by Viniciusap** [dotnet/csharplang#6815] [General] [answered]
Hello everyone!

Recently we were discussing one behavior on [dotnet/docs](https://github.com/dotnet/docs/) about this [issue](https://github.com/dotnet/docs/issues/33045).

I want to ask which approach we should commit to.

We find an error about the `optional parameters` not working well when we use them with another parameter using the `params` key.

You can see some examples below:

This example compiles but the Default value parameter is never used: [Example 1](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwUUBmA9gwBQBEGApgtADZWtwABOy69CrAJQBuekzYAPAcM7cofJSNU9x02Swx0eSYYqGbe/MyrGSZ5CvgAsgxi3xVigg0cEBeZRx0EKhGSgAOEMgQYADOgu7EANoAul6iajESFADeFIL58VQAnMwAJOyGxtnM3giCUHEwITyCAPzCAFRNPDwdrIIgXpUSAL62eQXuJeXWGYPZszwxAHQAMhwwAOYIABZjuuQFLgzIHBAAxjuCzABukfUIHGD1MGlamRNH+VNlrAA6CGyUEeYH2dhGQA==)

This forces the usage of the `optional parameter` but does not compile :  [Example 2](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwW3kUBmA9kwBQYMA2SABIXDwBEGAKYJonKoIHCxEwoICUAbgo91G9XXxUAbD3wAWHszawEPDtx4BeHlQEAHCMghgAzgarEA2gF1LOShOd0UKAG81DR0ATlYAEmEuXnD2ZJ4oTxhUTk4eAH4hACps3KLBHhBLZMUAXyVVck1YhNlxYPcq8NF2kIA6ABkRGABzBAALepUo9WZkEQgAY3GeVgA3FwyEETAMmEDe0JnNLzjEgB0EcKhtsCnG2qA==)

This forces the usage of the `optional parameter`  default value but does not compile :  [Example 3](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgAwAE+xALANwUUwRgCmAzgA4QBjTjQDCAexjcxAG04BBXrwoBvCjXV0AzHUJ1iAdjUbV5Ixo0B6KzQA83VGEjIAngD4z5mlYs0P577YW9o4QLu7knpbWNvzIHDTsXAC8AEShAObcKa42FrEc4ZHq+Nr0AGx0DDQAstAwABT0VADaALo0GdwAlH7GvZEAMlDcCDZNrjQIPAg0SQmcAO71PRFF5iZraykYnAjQ0sQpcP2b27v7hCkn5gC+LKarmzQAZmJi9VMjAHQAKmJyyDizmWXXuaxurEennKlReb3qsBmGGe0hmc2IcBo+TA3D0LXaOz2UGk3ROGyeegAnPUACTbFFIGjKerI1E0YYJVDSaQ0AD8NBSACoYFzpIKUjQQDRWQgujcUqDrhp6NS6YT9twpcp1cTuF8BpwYOkEAALeWKqFrV7ITiCE00eoAN1C7KmYHZMGl511Kwp0OIqpSAB0EMooG7zWCihCoTGbhoKEA=)

Using these examples, we came up with the following suggestions:

1 - Take off the possibility to have an optional parameter with other parameters using params. One or another should be used never both. Possibly creating a new compiler error to be triggered when that happens.

2 - Correct the behavior to accept an optional parameter and put them at the end (CS0231 must add an exception to the optional parameter). Today we have a rule that params should be the last one. We may add an exception for this rule when using an optional parameter

-or-

3 - Correct the behavior to accept an optional parameter to be understood with a param key (IDE0060 it's not thrown when the param is the last one). This is what's happening today. The optional parameter it's not being identified when we are trying to use it together with one parameter using the key parameters no matter the types we are using. So we need to correct the behavior to understand the parameter with default values, even if it's not the last one when it is together with another parameter params.

There was a discussion about but they never assigned the problem or even propose this as a bug. [Discussion](https://github.com/dotnet/csharplang/discussions/1458)

**Comment by 333fred** [comment, marked as answer]
@Viniciusap I'm not sure where the bug is, or what your proposed solutions mean. We're not going to change the language to remove the ability to mix optional parameters and params, that's been there since the feature was introduced and there are plenty of existing apis that use it (and you can even manually put the attributes on parameters to create things like optional parameters in the middle of a parameter list, which the computer understands even though you can't use the syntax directly). I certainly think there could be an analyzer, but this repo isn't the place for that, as Bill mentioned on your docs issue.
